midp/midp.js

```diff
@@ -1340,3 +1355,138 @@ addUnimplementedNative("com/sun/j2me/content/InvocationStore.get0.(Lcom/sun/j2me
 addUnimplementedNative("com/sun/j2me/content/InvocationStore.getByTid0.(Lcom/sun/j2me/content/InvocationImpl;II)I", 0);
 addUnimplementedNative("com/sun/j2me/content/InvocationStore.resetFlags0.(I)V");
 addUnimplementedNative("com/sun/j2me/content/AppProxy.midletIsRemoved.(ILjava/lang/String;)V");
+
+
+var isFullScreen = false;
+MIDP.setFullScreen = function(isFS) {
+    isFullScreen = isFS;
+    MIDP.updateCanvas();
+}
+
+MIDP.updateCanvas = function() {
+    var sidebar = document.getElementById("sidebar");
+    var header = document.getElementById("drawer").querySelector("header");
+    var canvas = MIDP.Context2D.canvas;
+    sidebar.style.display = header.style.display =
+        isFullScreen ? "none" : "block";
+    var headerHeight = isFullScreen ? 0 : header.offsetHeight;
+    canvas.height = MIDP.ScreenHeight - headerHeight;
+    canvas.dispatchEvent(new Event('canvasresize'));
+}
+
+var isVKVisible = false;
+var keyboardHeight = 0;
+var pendingShowNotify = false;
+var pendingHideNotify = false;
+var keyboardVisibilityListenerResolve;
+
+function onKeyboardShown(evt) {
+    if (isVKVisible) {
+        console.warn("keyboardShown but we thought keyboard was already visible!");
+    }
+
+    isVKVisible = true;
+
+    if (pendingHideNotify) {
+        pendingHideNotify = false;
+    } else if (keyboardVisibilityListenerResolve) {
+        keyboardVisibilityListenerResolve(1);
+        keyboardVisibilityListenerResolve = null;
+    } else {
+        pendingShowNotify = true;
+    }
+}
+
+function onKeyboardHidden(evt) {
+    if (!isVKVisible) {
+        console.warn("keyboardHidden but we thought the keyboard was already hidden!");
+    }
+
+    isVKVisible = false;
+
+    if (pendingShowNotify) {
+        pendingShowNotify = false;
+    } else if (keyboardVisibilityListenerResolve) {
+        keyboardVisibilityListenerResolve(0);
+        keyboardVisibilityListenerResolve = null;
+    } else {
+        pendingHideNotify = true;
+    }
+}
+
+function onWindowResize(evt) {
+    var shouldSendRotate = false;
+    if (window.innerWidth > MIDP.ScreenWidth) {
+        console.warn("Window grew beyond initial width!");
+        MIDP.ScreenWidth = window.innerWidth;
+        shouldSendRotate = true;
+    } else if (window.innerWidth < MIDP.ScreenWidth) {
+        console.warn("Window shrunk from initial width!");
+        MIDP.ScreenWidth = window.innerWidth;
+        shouldSendRotate = true;
+    }
+
+    if (window.innerHeight < MIDP.ScreenHeight) {
+        window.dispatchEvent(new Event("keyboardShown"));
```

Comment:
Rotation can cause window.innerHeight < MIDP.ScreenHeight. We cannot assert keyboard is visible in this case.